runtime/src/tests/functionalTests/page_objects/base.page.ts

```diff
@@ -0,0 +1,62 @@
+import { browser } from 'protractor';
+import * as mixins from '../mixins';
+import * as support from '../support';
+
+
+export enum PageOpenMode {
+  AlreadyOpened,
+  RefreshBrowser
+}
+
+export abstract class BasePage {
+  // add logging mixin
+
+  name: string = '...';
+  log: (action: string, ...msg: string[]) => void;
+  debug: (context: string, ...msg: string[]) => void;
+
+  // Use undefined to indicate the url has not been set
+  // It will use be in openInBrowser to throw error if the caller forgot
+  // to set the url. Need to do this because '' is a valid url and
+  // refers to the baseUrl
+
+  protected url: string|undefined;
+
+  constructor(url?: string) {
+    this.url = url;
+    this.debug(`url: '${url}'`);
+  }
+
+  async ready() {
+  }
+
+
+  async open(mode: PageOpenMode =  PageOpenMode.AlreadyOpened): Promise<BasePage> {
+
+    if (mode === PageOpenMode.RefreshBrowser) {
+      await this.openInBrowser();
+    }
+
+    await this.ready();
+    this.log('Opened');
+    return this;
+  }
+
+  async openInBrowser() {
+    if (this.url === undefined ) {
+      throw Error('Trying to open and undefined url');
```

Comment:
`s/and/an/i` or rather, `URL isn't defined; failed to open.`

Reply:
:+1: